gateway/handler.go

```diff
@@ -750,47 +771,89 @@ func debugStr(path string) string {
 	return q
 }
 
-// Resolve the provided contentPath including any special handling related to
-// the requested responseFormat. Returned ok flag indicates if gateway handler
-// should continue processing the request.
-func (i *handler) handlePathResolution(w http.ResponseWriter, r *http.Request, responseFormat string, contentPath ipath.Path, logger *zap.SugaredLogger) (resolvedPath ipath.Resolved, newContentPath ipath.Path, ok bool) {
-	// Attempt to resolve the provided path.
-	resolvedPath, err := i.api.ResolvePath(r.Context(), contentPath)
+func (i *handler) handleIfNoneMatch(w http.ResponseWriter, r *http.Request, responseFormat string, contentPath ipath.Path, imPath ImmutablePath, logger *zap.SugaredLogger) (ipath.Resolved, bool) {
+	// Detect when If-None-Match HTTP header allows returning HTTP 304 Not Modified
+	if inm := r.Header.Get("If-None-Match"); inm != "" {
+		// TODO: should we be handling _redirect and legacy 404 here?
+		gwMetadata, _, err := i.api.Head(r.Context(), imPath)
```

Comment:
> flagging we trigger additional read here that was not present in the old API.

Are you sure? Maybe I'm misunderstanding. `handleIfNoneMatch` returns the last CID + path after calling HEAD, which is then used later.

This is approximately the same behavior as the current incremental approach. There is maybe a little bit of extra fetching (e.g. for determining content type), which comes as part of using `Head` rather than `ResolvePath`.

I'm fine adding a `ResolvePath` API here if we want to though. WDYT?

Reply:
Added `ResolvePath`